apps/heft/src/configuration/HeftParameterManager.ts

```diff
@@ -0,0 +1,166 @@
+import { InternalError } from '@rushstack/node-core-library';
+import type { CommandLineParameter, CommandLineParameterProvider } from '@rushstack/ts-command-line';
+
+import type {} from '../pluginFramework/IHeftPlugin';
+import type {
+  HeftPluginDefinitionBase,
+  IChoiceParameterAlternativeJson,
+  IParameterJson
+} from './HeftPluginDefinition';
+
+export interface IParameterConfiguration {
+  pluginDefinition: HeftPluginDefinitionBase;
+  parameter: IParameterJson;
+}
+
+export class HeftParameterManager {
+  private _isFinalized: boolean = false;
+  private _parameterConfigurations: Set<IParameterConfiguration> = new Set();
+  // plugin definition => Map< parameter long name => applied parameter >
+  private _parametersByDefinition: Map<HeftPluginDefinitionBase, Map<string, CommandLineParameter>> =
+    new Map();
+
+  public addPluginParameters(pluginDefinition: HeftPluginDefinitionBase): void {
+    if (this._isFinalized) {
+      throw new InternalError('Parameters have already been finalized.');
+    }
+
+    // Parameters from this plugin have already been added
+    if (this._parametersByDefinition.has(pluginDefinition)) {
+      return;
+    } else {
+      this._parametersByDefinition.set(pluginDefinition, new Map());
+    }
+
+    // Add the parameters provided by the plugin. We will perform parameter conflict resolution when applying
+    for (const parameter of pluginDefinition.pluginParameters) {
+      this._parameterConfigurations.add({ parameter, pluginDefinition });
```

Comment:
This collection is redundant, since it can be derived at finalization time by iterating over the keys of `this._parametersByDefinition`, unless `pluginDefinition.pluginParameters` is mutable?
Would be faster to iterate that way as well, since the loop could reuse the value of the definition map for the specified plugin.

Reply:
You're right, pluginDefinition.pluginParameters is immutable, so yes they can be obtained from that.